nixos/containers: allow containers with long names to create private networks

Patryk27 opened the pull request:
###### Motivation for this change

Launching a container with a private network requires creating a
dedicated networking interface for it; name of that interface is derived
from the container name itself - e.g. a container named `foo` gets
attached to an interface named `ve-foo`.

An interface name can span up to IFNAMSIZ characters, which means that a
container name must contain at most IFNAMSIZ - 3 - 1 = 11 characters;
it's a limit that we validate using a build-time assertion.

This limit has been upgraded with Linux 5.8, as it allows for an
interface to contain a so-called altname, which can be much longer,
while remaining treated as a first-class citizen.

Since altnames have been supported natively by systemd for a while now,
due diligence on our side ends with dropping the name-assertion on newer
kernels.

This merge request closes #38509.

systemd/systemd#14467
systemd/systemd#17220
https://lwn.net/Articles/794289/

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Patryk27 commented:
TBH, I'm not sure why the pipeline failed :/

flokli commented:
As this only checks for the booted kernel (which can differ from the one currently running) how does this behave when switching to a new system with 5.8+ configured, but having booted an older kernel? Is there an error message printed by nspawn?

Or are the links just named differently?

Patryk27 commented:
Nice catch - in that case switching fails with:
```
machine # [    4.283293] container really-long-name[741]: Failed to set alternative interface name 've-really-long-name' to 've-really-l3QgY', ignoring: Operation not supported
machine # [    8.726324] container really-long-name[2492]: Cannot find device "ve-really-long-name"
```
Maybe our container-startup-script (the one that crashes with `Cannot find device` above) could check kernel's version too and print more fine-grained error message telling user to reboot the machine?

flokli commented:
Hmm - I wasn't entirely sure, just thinking out loudly.

If this is shown inside the activation script, it already gives quite a lot of information.

Maybe if we combine it with a 21.03 release note entry explaining the change, and a "make sure you're running a kernel with support for alternative interface names (5.8+)" at the end, that should be enough to connect the dots?

Patryk27 commented:
Release note seems alright for me :-)

Changing kernel versions requires users to make conscious effort (i.e. to modify some part of their configuration), so I consider `nixos-rebuild boot` instead of `nixos-rebuild switch` a natural order of things (and only doing `nixos-rebuild switch` with different kernel version can brake current approach).

Maybe we could add a warning (like `please don't forget to reboot`) for when someone does `nixos-rebuild switch` that changes kernel version?

tomberek commented:
Ran into this issue and would be nice to have this. What remains to be done/reviewed?

Patryk27 commented:
As far as I'm concerned, this feature is ready 🙂 

As @flokli noticed, we've got a small shortcoming that when you upgrade from an older kernel to 5.8+, the kernel version assertion _passes_ and thus Nix compiles the code which then fails at runtime during `nixos-rebuild switch` (because the system is still running on the older kernel); imho trying to somehow solve this is not worth the hassle, as upgrading the kernel is anyway a "big change" that almost always requires `reboot`, but feel free to add your 5 cents.

flokli commented:
The problem here is that even on latest `master`, `linuxPackages.kernel` points to 5.4.x, which is < 5.8+.

Patryk27 commented:
The new test already contains:
```nix
boot.kernelPackages = pkgs.linuxPackages_latest;    
```
... plus it _passes_ on my machine, so I wouldn't bet my money on an older kernel being loaded.

I've just rebased my branch and we'll see now - maybe it was just some spurious failure.

flokli commented:
Aaargh, sorry, I'm dumb. It's already too long ago - I forgot long names are only enabled on new kernels.

I locally cherry-picked the commit into master and successfully ran the test.

Can you:
 - Add the test to `nixos/tests/all-tests.nix`, so it appears in the `nixosTests` attrset
 - put the comment removal in a separate commit

flokli commented on nixos/tests/containers-names.nix:
```diff
@@ -0,0 +1,37 @@
+import ./make-test-python.nix ({ pkgs, ...} : {
+  name = "containers-names";
+  meta = with pkgs.stdenv.lib.maintainers; {
+    maintainers = [ patryk27 ];
+  };
```
```suggestion
import ./make-test-python.nix ({ pkgs, lib,  ...} : {
  name = "containers-names";
  meta = {
    maintainers = with lib.maintainers; [ patryk27 ];
  };
```

Don't use pkgs.lib, but `lib` directly.

Patryk27 replied:
Fix'd, thanks 🙁

Patryk27 committed:
nixos/containers: styling improvements

Patryk27 committed:
nixos/containers: allow containers with long names to create private networks

Launching a container with a private network requires creating a
dedicated networking interface for it; name of that interface is derived
from the container name itself - e.g. a container named `foo` gets
attached to an interface named `ve-foo`.

An interface name can span up to IFNAMSIZ characters, which means that a
container name must contain at most IFNAMSIZ - 3 - 1 = 11 characters;
it's a limit that we validate using a build-time assertion.

This limit has been upgraded with Linux 5.8, as it allows for an
interface to contain a so-called altname, which can be much longer,
while remaining treated as a first-class citizen.

Since altnames have been supported natively by systemd for a while now,
due diligence on our side ends with dropping the name-assertion on newer
kernels.

This commit closes #38509.

systemd/systemd#14467
systemd/systemd#17220
https://lwn.net/Articles/794289/

flokli commented:
Thanks!
